Updated the economy-system

jateute opened the pull request:
Added user-feedback and fixed a few bugs

SCDerox commented:
Thanks so much for your contribution - to make review easier, could you provide us with a list of (User facing) changes? Doesn't need to be detailed, but should include the most important changes.

jateute commented:
Changes:
- Added ids for items
- Added a select menu for the shop
- Added a shop-message (overview over the shop; gets send in a channel and works like the leaderboard)

SCDerox commented on modules/economy-system/events/botReady.js:
```diff
@@ -26,6 +26,21 @@ module.exports.run = async function (client) {
+        client.logger.info('[economy-system] ' + localize('economy-system', 'migration-happening'));
+        const data = await client.models['economy-system']['Shop'].findAll({attributes: ['name', 'price', 'role']});
+        await client.models['economy-system']['Shop'].sync({force: true});
+        i = 0;
```
```suggestion
        let i = 0;
```
variable is only declared implicitly

SCDerox commented on modules/economy-system/economy-system.js:
```diff
@@ -124,67 +125,274 @@ async function editBank(client, id, action, value) {
+            placeholder: localize('economy-system', 'nothing-selected'),
+            'min_values': 1,
+            'max_values': 1,
+            options: options,
```
This is an empty array if no items are in the shop, causing a crash on message send

SCDerox commented on modules/economy-system/commands/shop.js:
```diff
@@ -1,92 +1,42 @@
-            u: interaction.user.tag,
-            i: item['value']
-        }));
+    'add': async function (interaction) { // ToDo: Update Command
+        if (!checkPerms(interaction)) return;
+        await createShopItem(interaction);
+        shopMsg(interaction.client);
```
(so there is a todo left here…)
Also: For better UX, I would suggest checking if the bot's highest role is above the supplied role

SCDerox reviewed (changes requested):
Thanks so much for your contribution - here are some issues I run into when trying to test your module

SCDerox commented:
Do you have any updates on this?

jateute commented:
> Do you have any updates on this?

Unfortunately I hadn‘t any time to finish the fixes. I‘ll try to resolve the problems asap

SCDerox commented on modules/economy-system/economy-system.js:
```diff
@@ -124,67 +125,281 @@ async function editBank(client, id, action, value) {
     for (let i = 0; i < items.length; i++) {
-        string = `${string}**${items[i].dataValues.name}**: ${items[i].dataValues.price} ${client.configurations['economy-system']['config']['currencySymbol']}\n`;
+        const roles = await guild.roles.fetch(items[i].dataValues.role);
+        string = `${string}${inputReplacer({'%id%': items[i].dataValues.id, '%itemName%': items[i].dataValues.name, '%price%': `${items[i].dataValues.price} ${client.configurations['economy-system']['config']['currencySymbol']}`, '%sellcount%': roles.members.size}, client.configurations['economy-system']['strings']['itemString'])}`;
```
If the role got deleted, this causes a crash

SCDerox commented on modules/economy-system/commands/shop.js:
```diff
@@ -1,92 +1,42 @@
-        }));
+        let name = await interaction.options.get('item-name');
+        if (name) name = name['value'];
+        let id = await interaction.options.get('item-id')['value'];
```
This will cause a crash if item-id is not set. Use [`CommandInteractionOptionResolver.getString()`](https://discord.js.org/#/docs/discord.js/v13/class/CommandInteractionOptionResolver?scrollTo=getString) to simplify this

SCDerox commented on modules/economy-system/commands/shop.js:
```diff
@@ -1,92 +1,42 @@
+        const msg = await createShopMsg(interaction.client, interaction.guild, !interaction.client.configurations['economy-system']['config']['publicCommandReplies']);
         interaction.reply(msg);
     },
     'delete': async function (interaction) {
```
This command will cause crash if no parameters were set

jateute replied:
This command doesn't have any parameters. Or what do you mean by "if no parameters were set"?

SCDerox replied:
I think I selected something wrong here… What I meant by that is a crash if users try to /shop buy an item, but do not set any parameters in the command (I think, but I'm not sure myself so…)

SCDerox reviewed (changes requested):
I've found some critical edge-cases causing a crash on command execution. In addition, I fixed some small typos :sweat_smile:. Please review my comments.  
In addition, I would suggest auto-complete support for both the ID and name field (this could reduce the two fields to one, if you implement search similarly to the way used in the [polls module](https://github.com/ScootKit/CustomDCBot/blob/d099bffea3b61f5e98f19fc125f1cfffbf10b6e8/modules/polls/commands/poll.js#L49)).
Additionally, the shop message does not seem to get send into the configured channel. could you check that?
Apart form these issues, your changes seem to work fine and are ready to be merged, once these issues got resolved.

SCDerox commented:
Do you have an update on this? :sweat_smile:

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ScootKit/CustomDCBot?pullRequest=98) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ScootKit/CustomDCBot?pullRequest=98) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ScootKit/CustomDCBot?pullRequest=98) it.</sub>

ElBe-Plaq commented on modules/economy-system/configs/config.json:
In [line 300](https://github.com/jateute/CustomDCBot/blob/main/modules/economy-system/configs/config.json#L300) there is a grammatical error. Instead of "Chance to **become** money for a message (Chance: 1/ This value). Set to 0 to disable message drop" it should be "Chance to **get** money for a message (Chance: 1/ This value). Set to 0 to disable message drop".

ElBe-Plaq commented:
Also, when `/economy crime` fails, the money you loose is calculated and taken from your [deposited money, instead of your bank money](https://github.com/jateute/CustomDCBot/blob/main/modules/economy-system/commands/economy-system.js#L62-L63). A possible solution to this would be to take it from where the most money is (take money from bank if there is more, else take it from deposited money).

ElBe-Plaq commented:
It really needs help...

![image](https://github.com/ScootKit/CustomDCBot/assets/90863907/9eccc830-d94d-49c9-b81c-4e67f8c8f41a)

Translation: 
### Balance of **elbe**#0

**Wallet:**
null 🪙

**Bank:**
null 🪙

**All:**
NaN 🪙

SCDerox commented:
@jateute I have converted this PR to a draft to make it more clear that it isn't ready to be used and work in progress. You can convert it back once it's ready to use.

jateute commented:
> @jateute I have converted this PR to a draft to make it more clear that it isn't ready to be used and work in progress. You can convert it back once it's ready to use.

Thx, I didn't thought of it :hAA:

ElBe-Plaq commented:
> Also, when `/economy crime` fails, the money you loose is calculated and taken from your [deposited money, instead of your bank money](https://github.com/jateute/CustomDCBot/blob/main/modules/economy-system/commands/economy-system.js#L62-L63). A possible solution to this would be to take it from where the most money is (take money from bank if there is more, else take it from deposited money).
> 
> (Sadly the screenshot is only available in German)
> ![image](https://github.com/ScootKit/CustomDCBot/assets/90863907/9a4d3700-15b5-40d4-b819-d5205b0df80a)
> 
> Translation: "You tried to steal a wallet but got caught. You lose **0 🪙**."

It could also work by giving you debt, so you don't lose anything and don't have 0 coins, but rather -X coins.

jateute commented:
> It really needs help...
> 
> ![image](https://private-user-images.githubusercontent.com/90863907/295174034-9eccc830-d94d-49c9-b81c-4e67f8c8f41a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzMjc4NDgsIm5iZiI6MTcxOTMyNzU0OCwicGF0aCI6Ii85MDg2MzkwNy8yOTUxNzQwMzQtOWVjY2M4MzAtZDk0ZC00OWM5LWI4MWMtNGU2N2Y4YzhmNDFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI1VDE0NTkwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYxNDI0Yjc5MjMwOGU4YTUxNzkzOWUxNjU4MmIyMjg2OGFlNWMwMjI4YzkwYjgxM2Y4NGVmOTVkYzk1YTRkZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xFC62pIgkPA68xR0xitocriV2rYXQLtsyEJEXNAtBRY)
> 
> Translation:
> ### Balance of **elbe**#0
> 
> **Wallet:** null 🪙
> 
> **Bank:** null 🪙
> 
> **All:** NaN 🪙

How did you do that? I wasn't able to reproduce that bug. For me the user either doesn't exist or I get normal numerical values

ElBe-Plaq commented:
> > It really needs help...
> > 
> > ![image](https://private-user-images.githubusercontent.com/90863907/295174034-9eccc830-d94d-49c9-b81c-4e67f8c8f41a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzMjc4NDgsIm5iZiI6MTcxOTMyNzU0OCwicGF0aCI6Ii85MDg2MzkwNy8yOTUxNzQwMzQtOWVjY2M4MzAtZDk0ZC00OWM5LWI4MWMtNGU2N2Y4YzhmNDFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI1VDE0NTkwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYxNDI0Yjc5MjMwOGU4YTUxNzkzOWUxNjU4MmIyMjg2OGFlNWMwMjI4YzkwYjgxM2Y4NGVmOTVkYzk1YTRkZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xFC62pIgkPA68xR0xitocriV2rYXQLtsyEJEXNAtBRY)
> > 
> > Translation:
> > ### Balance of **elbe**#0
> > 
> > **Wallet:** null 🪙
> > 
> > **Bank:** null 🪙
> > 
> > **All:** NaN 🪙
> 
> How did you do that? I wasn't able to reproduce that bug. For me the user either doesn't exist or I get normal numerical values

I somehow was able to get negative money I think. Sadly the server this screenshot was taken on is deleted now, so I can't check exactly what I did.

jateute commented:
> > > It really needs help...
> > > ![image](https://private-user-images.githubusercontent.com/90863907/295174034-9eccc830-d94d-49c9-b81c-4e67f8c8f41a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzMjc4NDgsIm5iZiI6MTcxOTMyNzU0OCwicGF0aCI6Ii85MDg2MzkwNy8yOTUxNzQwMzQtOWVjY2M4MzAtZDk0ZC00OWM5LWI4MWMtNGU2N2Y4YzhmNDFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI1VDE0NTkwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYxNDI0Yjc5MjMwOGU4YTUxNzkzOWUxNjU4MmIyMjg2OGFlNWMwMjI4YzkwYjgxM2Y4NGVmOTVkYzk1YTRkZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xFC62pIgkPA68xR0xitocriV2rYXQLtsyEJEXNAtBRY)
> > > Translation:
> > > ### Balance of **elbe**#0
> > > **Wallet:** null 🪙
> > > **Bank:** null 🪙
> > > **All:** NaN 🪙
> > 
> > 
> > How did you do that? I wasn't able to reproduce that bug. For me the user either doesn't exist or I get normal numerical values
> 
> I somehow was able to get negative money I think. Sadly the server this screenshot was taken on is deleted now, so I can't check exactly what I did.

![image](https://github.com/ScootKit/CustomDCBot/assets/78431813/c9abd009-d747-4ffb-ab23-974e7f193229)
For me it also works with negative numbers :) Maybe I somehow fixed it

SCDerox commented on modules/economy-system/configs/strings.json:
```diff
@@ -518,6 +581,106 @@
+      },
+      "default": {
+        "en": "You already own this Item",
+        "de": "Du hast diese Item bereits gekauft"
```
```suggestion
        "de": "Du hast dieses Item bereits gekauft"
```

SCDerox commented on modules/economy-system/configs/strings.json:
```diff
@@ -518,6 +581,106 @@
+        "de": "Du hast diese Item bereits gekauft"
+      },
+      "description": {
+        "en": "The message that is send when the user trys to buy an Item that he aleady own",
```
```suggestion
        "en": "The message that is send when the user trys to buy an Item that he already own",
```

SCDerox reviewed (changes requested):
Thanks so much for submitting this PR for review - I apologize for this delay, I've overlocked your awesome PR.

Please note that in my review, I've added some comments that are not required for the merge of this PR, but are suggestions. These contain the key word "consider". You are welcome to comment / dismiss them, if you do not agree with them. 
Please review all other comments carefully and either implement them or comment on them, so we can come to a decision together. 


* I've tested featured mentioned in [this command](https://github.com/ScootKit/CustomDCBot/pull/98#issuecomment-1369583241). If you've added additional features, feel free to comment, so I can get that noted internally for our changelogs.
* Users can rob themselves. 
* Consider adding a command that allows shop owners to view an item attributes (price, assigned role, …).
* Consider moving all shop owner commands to a separate command, such as `/shopadmin`. This would increase UX as users won't see commands that they can't use (also, you can take advantage of Discord's slash command permission handling)

jateute commented on modules/economy-system/economy-system.js:
```diff
@@ -124,86 +124,317 @@ async function editBank(client, id, action, value) {
+    });
+    await editBalance(interaction.client, interaction.user.id, 'remove', item[0]['price']);
+    console.log(item[0]['role']);
+    await interaction.member.roles.add(item[0]['role']);
```
I did this so that the user definitely has enough money to buy the item. And I check the perms when creating the item. Therefor it could only lead to errors if the server owner or some admin changes the perms afterwards

SCDerox replied:
I was aware that permission checks are being made before the item gets created, but admins often change role positions, permissions and more, sometimes even deleting the role. Adequate handling of such cases is required, as users get charged for a role that they might not receive. The opposite case, where a user loses money after the initial balance check has already been completed, but the role not yet assigned (a race-condition) is more unlikely, but if you want to handle such cases, please add a system that refunds the charge if role assignment fails (adding to balance when the role has not been assigned)

jateute commented on modules/economy-system/economy-system.js:
```diff
@@ -244,7 +475,11 @@ async function leaderboard(client) {
 module.exports.deleteShopItem = deleteShopItem;
 module.exports.createShopMsg = createShopMsg;
+module.exports.shopMsg = shopMsg;
 module.exports.createleaderboard = leaderboard;
```
Is there any particular reason that you resolved the conversation? @SCDerox

SCDerox replied:
I don't think I resolved this conversation :sweat_smile:

jateute replied:
It somehow was marked as resolved by you :hAA: I unresolved it ^^ I was just confused

jateute commented:
It seems like I forgot to push the changes. I‘ll check on it later, when I‘m back at the PC ^^

SCDerox commented on modules/economy-system/commands/shop.js:
```diff
@@ -1,83 +1,55 @@
-            i: item['value']
-        }));
+        if (!checkPerms(interaction)) return;
+        interaction.reply({
+            content: localize('economy-system', 'creating-item'),
+            ephemeral: true
+        });
```
```suggestion
        interaction.reply({
            content: localize('economy-system', 'creating-item'),
            ephemeral: true
        });
```
This creates a bunch of issues. What you are likely looking for is first the `await interaction.deferReply({ephemeral: true})` call, after which all `interaction.reply` calls would need to be replaced by `editReply` calls to avoid issues (same with other affected commands, such as /shop delete`.

SCDerox commented on modules/economy-system/commands/shop.js:
```diff
@@ -1,83 +1,47 @@
+        await createShopItem(interaction);
+        await shopMsg(interaction.client);
     },
     'buy': async function (interaction) {
```
No deferred reply was added to this command in particular

SCDerox reviewed (approved):
Thank you for your patience and contribution, everything looks fine to me and works as expected.
I'll go ahead and merge your PR. I'm excited to announce that your updates to your module will be available on SCNX shortly 🚀
